clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java

```diff
@@ -1210,9 +1210,11 @@ public void assign(Collection<TopicPartition> partitions) {
                 client.maybeTriggerWakeup();
 
                 if (includeMetadataInTimeout) {
-                    if (!updateAssignmentMetadataIfNeeded(timer)) {
-                        return ConsumerRecords.empty();
-                    }
+                    // try to update assignment metadata BUT do not need to block on the timer,
+                    // since even if we are 1) in the middle of a rebalance or 2) have partitions
+                    // with unknown starting positions we may still want to return some data
+                    // as long as there are some partitions fetchable
+                    updateAssignmentMetadataIfNeeded(time.timer(1L));
```

Comment:
This is the tricky part: I have to use a non-zero timer so that blocking rpc like find-coordinator is guaranteed to poll for once, otherwise the `future.hasExpired` will trigger immediately and we would doomed to be not finishing that call.

Reply:
I am wondering about two things here:
- How robust is this solution? Is it guaranteed that with the timer set to 1 the call is done? 
- If I look at https://github.com/apache/kafka/blob/d112ffd8d85de2441a8ea32d080e584d7d16a9a5/clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkClient.java#L210-L215 the poll is done also if the timer is set to zero. Do I look at the wrong code? If yes, where is the code you refer to?

Reply:
If the timer is set to zero, it means that we would not send the request at all -- note that `client.send` call just queued up the request in the queue, and only `client.poll` would write it to the socket, and that's why I cannot have the timer to be 0.

Setting it to be 1 should be robust since the `timer` is initialized at the beginning of the call and until the first `client.poll()` it would not be checked, and hence not returned early.

Reply:
I've further polished it to only use `timer(1)` if the remaining is still > 0, in this case we still make sure that `consumer.poll(0)` can return instantaneously.